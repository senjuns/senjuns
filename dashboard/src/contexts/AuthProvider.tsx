import { Amplify, Auth } from 'aws-amplify';

import {
  FC,
  createContext,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useState,
} from 'react';
import { useHistory } from 'react-router-dom';

import { APP_URL } from '../shared/constants';

/**
 * AuthContext definition
 */
export interface AuthContextType {
  /**
   * Shows whether the user is logged in or not.
   */
  isLoggedIn: boolean;
  /**
   * Callback to login with email and password.
   */
  logIn: (email: string, password: string) => void;
  /**
   * Callback to sign out from the application.
   */
  signOut: () => Promise<void>;
  /**
   * UserInfo that has the authenticated user information.
   */
  userInfo: any;
}

export const AuthContext = createContext<AuthContextType>(undefined!);

export const AuthProvider: FC<any> = ({ children }) => {
  const history = useHistory();
  const [userInfo, setUserInfo] = useState<any>();
  const [isLoggedIn, setIsLoggedIn] = useState(false);

  const authenticateCurrentUser = useCallback(async () => {
    // eslint-disable-next-line @typescript-eslint/no-shadow
    const userInfo = await Auth.currentAuthenticatedUser();
    setUserInfo(userInfo);
    setIsLoggedIn(true);
  }, []);

  const signOut = useCallback(async () => {
    await Auth.signOut();
    setIsLoggedIn(false);
    history.push(APP_URL.login);
  }, [history]);

  const logIn = useCallback(async (email: string, password: string) => {
    // eslint-disable-next-line @typescript-eslint/no-shadow
    const userInfo = await Auth.signIn(email, password);
    setUserInfo(userInfo);
    setIsLoggedIn(true);
  }, []);

  useEffect(() => {
    void authenticateCurrentUser();
  }, [authenticateCurrentUser]);

  useEffect(() => {
    fetch('/runtime-config.json')
      .then((response) => response.json())
      .then((runtimeContext) => {
        runtimeContext.region &&
          runtimeContext.userPoolId &&
          runtimeContext.userPoolWebClientId &&
          runtimeContext.identityPoolId &&
          Amplify.configure({
            aws_appsync_graphqlEndpoint: runtimeContext.appSyncGraphqlEndpoint,
            Auth: {
              region: runtimeContext.region,
              userPoolId: runtimeContext.userPoolId,
              userPoolWebClientId: runtimeContext.userPoolWebClientId,
              identityPoolId: runtimeContext.identityPoolId,
            },
          });
      })
      .catch((e) => console.log(e));
  }, []);

  const values = useMemo<AuthContextType>(
    () => ({
      isLoggedIn,
      logIn,
      signOut,
      userInfo,
    }),
    [isLoggedIn, logIn, signOut, userInfo],
  );

  return <AuthContext.Provider value={values}>{children}</AuthContext.Provider>;
};

export const useAuth = () => {
  return useContext(AuthContext);
};
