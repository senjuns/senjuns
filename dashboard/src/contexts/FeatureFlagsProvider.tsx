import Cookies from 'js-cookie';
import { createContext, FC, useContext, useEffect, useState } from 'react';
import { FEATURE_ENABLED } from '../shared/constants';
import { FEATURE_FLAGS } from '../shared/interfaces';

export interface FeatureFlagsContextState {
  heatMapViewFlag: boolean;
  settingsFlag: boolean;
  dashboardFlag: boolean;
  lineChartFlag: boolean;
  calendarFlag: boolean;
  notificationFlag: boolean;
  debugFlag: boolean;
  internalHeatmapToolFlag: boolean;
  airflowSensorFlag: boolean;
}

export const FeatureFlagsContext = createContext<FeatureFlagsContextState>(
  undefined!,
);

/**
 * This is the provider component that provides the Feature Flags
 * The feature flags can be fetched from the backend or can be stored in the cookie or whatever.
 * @param {FC} props The props
 * @returns {JSX.Element} The rendered element.
 */
export const FeatureFlagsProvider: FC<any> = (props) => {
  const [heatMapViewFlag, setHeatMapViewFlag] = useState(true);
  const [settingsFlag, setSettingsFlag] = useState(false);
  const [dashboardFlag, setDashboardFlag] = useState(false);
  const [lineChartFlag, setLineChartFlag] = useState(false);
  const [calendarFlag, setCalendarFlag] = useState(false);
  const [notificationFlag, setNotificationFlag] = useState(false);
  const [debugFlag, setDebugFlag] = useState(false);
  const [internalHeatmapToolFlag, setInternalHeatmapToolFlag] = useState(false);
  const [airflowSensorFlag, setAirflowSensorFlag] = useState(false);

  useEffect(() => {
    setHeatMapViewFlag(true);
    setSettingsFlag(Cookies.get(FEATURE_FLAGS.SETTINGS) === FEATURE_ENABLED);
    setDashboardFlag(Cookies.get(FEATURE_FLAGS.DASHBOARD) === FEATURE_ENABLED);
    setLineChartFlag(Cookies.get(FEATURE_FLAGS.LINE_CHART) === FEATURE_ENABLED);
    setCalendarFlag(Cookies.get(FEATURE_FLAGS.CALENDAR) === FEATURE_ENABLED);
    setNotificationFlag(
      Cookies.get(FEATURE_FLAGS.NOTIFICATION) === FEATURE_ENABLED,
    );
    setDebugFlag(Cookies.get(FEATURE_FLAGS.DEBUG) === FEATURE_ENABLED);
    // cookieStore.set("FEATURE_INTERNAL_HEATMAP_TOOL_FLAG", "FEATURE_ENABLED");
    setInternalHeatmapToolFlag(
      Cookies.get(FEATURE_FLAGS.INTERNAL_HEATMAP_TOOL_FLAG) === FEATURE_ENABLED,
    );
    setAirflowSensorFlag(
      Cookies.get(FEATURE_FLAGS.AIRFLOW_SENSOR_FLAG) === FEATURE_ENABLED,
    );
  }, []);

  return (
    <FeatureFlagsContext.Provider
      {...props}
      value={{
        heatMapViewFlag,
        settingsFlag,
        dashboardFlag,
        lineChartFlag,
        calendarFlag,
        notificationFlag,
        debugFlag,
        internalHeatmapToolFlag,
        airflowSensorFlag,
      }}
    />
  );
};

export const useFeatureFlags = () => {
  return useContext(FeatureFlagsContext);
};
