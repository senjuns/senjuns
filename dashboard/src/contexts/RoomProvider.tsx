import React, { FC, useEffect, useMemo, useState, useContext } from 'react';
import { useHistory } from 'react-router-dom';
import { DropdownOption } from '../components/common';
import {
  GetZonesByOrganizationCodeQuery,
  useGetSystemsByZoneIdQuery,
  useGetZonesByOrganizationCodeQuery,
} from '../graphql/generated/react_apollo';
import { useScreenSize } from '../hooks';
import { ROOM_DETAIL_TABS, NULL_VALUE } from '../shared/constants';
import { getCurrentDateTime } from '../shared/utils';

export interface RoomContextState {
  roomOptions: DropdownOption[];
  currentRoomId: number;
  setCurrentRoomId: (id: number) => void;
  currentTab: ROOM_DETAIL_TABS;
  setCurrentTab: (tab: ROOM_DETAIL_TABS) => void;
  currentSystemId: number;
  setCurrentSystemId: (id: number) => void;
  currentDateTime: Date;
  setCurrentDateTime: (date: Date) => void;
  isPhotoSectionView: boolean;
  setIsPhotoSectionView: (isView: boolean) => void;
}

export const RoomContext = React.createContext<RoomContextState>(undefined!);

/**
 * This component represents a provider that provides data fetched from the backend.
 * This is essentially global state used in the application.
 * *
 * @param {FC} props The props
 * @returns {JSX.Element} global state
 */
export const RoomProvider: FC = (props) => {
  const history = useHistory();
  const { isMobile } = useScreenSize();
  const [rooms, setRooms] = useState<
    GetZonesByOrganizationCodeQuery['zone'] | null
  >(null);

  // TODO: get org code from user
  const organizationCode = 'NEATLEAF';
  const organizationFilter = {
    variables: { code: { _eq: organizationCode } },
  };
  const { data } = useGetZonesByOrganizationCodeQuery(organizationFilter);

  useEffect(() => {
    setRooms(data?.zone || null);
  }, [data]);

  const [currentRoomId, setCurrentRoomId] = useState(NULL_VALUE);
  const [currentTab, setCurrentTab] = useState(ROOM_DETAIL_TABS.OVERVIEW);
  const [isPhotoSectionView, setIsPhotoSectionView] = useState(false);
  const [currentSystemId, setCurrentSystemId] = useState(NULL_VALUE);
  const [currentDateTime, setCurrentDateTime] = useState(getCurrentDateTime());

  useEffect(() => {
    if (isMobile) {
      setCurrentTab(ROOM_DETAIL_TABS.OVERVIEW);
    } else {
      setCurrentTab(ROOM_DETAIL_TABS.HEAT_MAP);
    }
  }, [isMobile]);

  const zoneFilter = {
    variables: {
      zone_id: currentRoomId,
    },
  };
  const { data: systemData } = useGetSystemsByZoneIdQuery(zoneFilter);

  useEffect(() => {
    setCurrentSystemId(systemData?.system?.[0]?.id || NULL_VALUE);
    setCurrentDateTime(getCurrentDateTime());
  }, [systemData?.system?.[0]]);

  useEffect(() => {
    setCurrentRoomId(rooms?.[0].id || NULL_VALUE);
  }, [rooms]);

  useEffect(() => {
    if (currentRoomId === NULL_VALUE) return;
    history.push(`/room-details/${currentRoomId}/${currentTab}`);
  }, [history, currentTab, currentRoomId]);

  const roomOptions: DropdownOption[] = useMemo(
    () =>
      (rooms || []).map((room) => ({
        id: room.id,
        label: room.enumeration.description,
      })),
    [rooms]
  );

  return (
    <RoomContext.Provider
      {...props}
      value={{
        roomOptions,
        currentRoomId,
        setCurrentRoomId,
        currentSystemId,
        setCurrentSystemId,
        currentTab,
        setCurrentTab,
        currentDateTime,
        setCurrentDateTime,
        isPhotoSectionView,
        setIsPhotoSectionView,
      }}
    />
  );
};

export const useRoomDetails = () => {
  return useContext(RoomContext);
};
