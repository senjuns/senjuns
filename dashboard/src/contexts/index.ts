export * from './AuthProvider';
export * from './FeatureFlagsProvider';
