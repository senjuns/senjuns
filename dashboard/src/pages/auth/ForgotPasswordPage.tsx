import { Auth } from 'aws-amplify';
import React, { useState } from 'react';
import { useForm } from 'react-hook-form';
import { useHistory } from 'react-router-dom';

import { NonAuthForm, NonAuthLayout } from '../../components/auth';
import {
  Button,
  Grid,
  Link,
  TextInput,
  Typography,
} from '../../components/common';
import { useScreenSize } from '../../hooks/useScreenSize';
import { APP_URL, AWS_API_CODE } from '../../shared/constants';

/**
 * ForgotPasswordPage component.
 *
 * @returns {JSX.Element} The rendered component.
 */
const ForgotPasswordPage: React.FC = () => {
  const [error, setError] = useState('');
  const [isSubmitting, setIsSubmitting] = useState(false);
  const { isMobile } = useScreenSize();

  const history = useHistory();
  const { handleSubmit, register } = useForm<FormData>();

  const onSubmit = handleSubmit(async (values: any) => {
    setIsSubmitting(true);

    try {
      await Auth.forgotPassword(values.email);
      history.push(APP_URL.resetPassword(values.email));
      // eslint-disable-next-line @typescript-eslint/no-shadow
    } catch (error) {
      if ((error as any).code === AWS_API_CODE.user_not_found_exception) {
        setError('There was error during forgot password');
      } else {
        setError('Server is unreachable');
      }
    } finally {
      setIsSubmitting(false);
    }
  });

  return (
    <NonAuthLayout>
      <NonAuthForm
        instructionHeader={
          <>
            <Typography fontWeight="normal" variant={isMobile ? 'h3' : 'h2'}>
              Forgot
            </Typography>
            <Typography variant={isMobile ? 'h3' : 'h2'}>password</Typography>
          </>
        }
        instructions={
          <Typography color="gray" lineHeight="30px" variant="body1">
            Enter your registered email below to receive password reset
            instructions
          </Typography>
        }
        footer={
          <>
            <Typography variant="body2">Remember password?</Typography>
            <Link to={APP_URL.login} text="Login" />
          </>
        }
      >
        <form onSubmit={onSubmit}>
          <Grid container spacing={3}>
            <Grid item xs={12}>
              <TextInput
                fullWidth
                label="Email"
                {...register('email' as any, { required: true })}
              />
            </Grid>

            <Grid item xs={12}>
              <Button disabled={isSubmitting} fullWidth type="submit">
                Send code
              </Button>
            </Grid>

            {error && (
              <Grid item xs={12}>
                {error}
              </Grid>
            )}
          </Grid>
        </form>
      </NonAuthForm>
    </NonAuthLayout>
  );
};

export default ForgotPasswordPage;
