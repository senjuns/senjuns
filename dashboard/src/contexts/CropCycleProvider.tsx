import { subDays } from 'date-fns';
import {
  createContext,
  FC,
  useContext,
  useState,
  useMemo,
  useEffect,
} from 'react';

import { useRoomDetails } from 'contexts/RoomProvider';
import { TimeRange } from 'shared/interfaces/general';
import { IGrowCycleInfo, useGrowCycle } from 'hooks/useGrowCycle';

export interface CropCycleState extends IGrowCycleInfo {
  range: TimeRange;
  onChange: (range: TimeRange) => void;
}

export const CropCycleContext = createContext<CropCycleState>(undefined!);

export const CropCycleProvider: FC = (props) => {
  const { currentRoomId, currentDateTime } = useRoomDetails();

  const {
    startTime,
    endTime,
    lightInfo,
    loading,
    monitoringParams,
    plantInfo,
  } = useGrowCycle(currentRoomId, currentDateTime);

  const defaultRange = useMemo(
    () => ({
      start: subDays(currentDateTime, 3),
      end: currentDateTime,
    }),
    [currentDateTime]
  );

  const [range, setRange] = useState<TimeRange>(defaultRange);

  useEffect(() => {
    setRange(defaultRange);
  }, [defaultRange]);

  return (
    <CropCycleContext.Provider
      {...props}
      value={{
        startTime,
        endTime,
        plantInfo,
        lightInfo,
        loading,
        monitoringParams,
        range,
        onChange: setRange,
      }}
    />
  );
};

export const useCropCycleDetails = () => {
  return useContext(CropCycleContext);
};
