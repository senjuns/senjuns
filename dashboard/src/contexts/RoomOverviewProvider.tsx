import MomentUtils from '@date-io/moment';
import React, { useContext, useEffect, useState } from 'react';
import { useGetLatestMeasurementRunBySystemIdQuery } from '../graphql/generated/react_apollo';
import { useCropCycleDetails } from './CropCycleProvider';
import { useRoomDetails } from './RoomProvider';

const momentUtils = new MomentUtils();

export interface RoomOverviewContextState {
  loading: boolean;
  statistics: any;
  updatedTime: string | null;
}

export const RoomOverviewContext =
  React.createContext<RoomOverviewContextState>(undefined!);

export const RoomOverviewProvider: React.FC = ({ children }) => {
  const [statistics, setStatistics] = useState<any>(null);
  const [updatedTime, setUpdatedTime] = useState<string | null>(null);

  const { currentSystemId, setCurrentDateTime } = useRoomDetails();

  const { range } = useCropCycleDetails();

  const { data, loading } = useGetLatestMeasurementRunBySystemIdQuery({
    variables: {
      system_id: { _eq: currentSystemId },
      start: range.start,
      end: range.end,
    },
  });

  useEffect(() => {
    const firstMeasurementRun = data?.measurement_run[0];
    if (firstMeasurementRun) {
      if (!firstMeasurementRun?.metadata?.statistics) {
        setCurrentDateTime(firstMeasurementRun.start_time);
      }

      setStatistics(firstMeasurementRun?.metadata?.statistics);
      const measurementRunTime = new Date(firstMeasurementRun?.start_time);
      setUpdatedTime(momentUtils.moment(measurementRunTime).from(range.start));
    } else {
      setStatistics(null);
      setUpdatedTime(null);
    }
  }, [data]);

  return (
    <RoomOverviewContext.Provider
      value={{
        loading,
        statistics,
        updatedTime,
      }}
    >
      {children}
    </RoomOverviewContext.Provider>
  );
};

export const useRoomOverview = () => {
  return useContext(RoomOverviewContext);
};
