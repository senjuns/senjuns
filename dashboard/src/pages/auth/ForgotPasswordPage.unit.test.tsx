import * as ShallowRenderer from 'react-test-renderer/shallow';
import { AuthProvider } from 'contexts';
import ForgotPasswordPage from './ForgotPasswordPage';

describe('<ForgotPasswordPage />', () => {
  it('renders correctly.', async () => {
    const renderer = ShallowRenderer.createRenderer();
    renderer.render(
      <AuthProvider>
        <ForgotPasswordPage />
      </AuthProvider>
    );

    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});
