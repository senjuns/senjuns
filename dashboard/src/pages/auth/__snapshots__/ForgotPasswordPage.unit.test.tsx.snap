// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ForgotPasswordPage /> renders correctly. 1`] = `
<Context.Provider
  value={
    Object {
      "isLoggedIn": false,
      "logIn": [Function],
      "signOut": [Function],
      "userInfo": undefined,
    }
  }
>
  <ForgotPasswordPage />
</Context.Provider>
`;
